checker/cost.go

```diff
@@ -0,0 +1,536 @@
+		sum.Add(selectCost)
+	case kindTypeParam:
+		// Set the operand type to DYN to prevent assignment to a potentionally inco
+		//rrect type
+		// at a later point in type-checking. The isAssignable call will update the type
+		// substitutions for the type param under the covers.
+		c.isAssignable(decls.Dyn, targetType)
+		// Also, set the result type to DYN.
+		resultType = decls.Dyn
+		sum.Add(selectCost)
+	default:
+		// Dynamic / error values are treated as DYN type. Errors are handled this way as well
+		// in order to allow forward progress on the cost.
+		if isDynOrError(targetType) {
+			resultType = decls.Dyn
+		} else {
+			// does not support field selection
+		}
+	}
+	if sel.TestOnly {
+		resultType = decls.Bool
+	}
+	c.setType(e, resultType)
+
+	return sum
+}
+
+func (c *coster) costCall(e *exprpb.Expr) CostEstimate {
+	call := e.GetCallExpr()
+	target := call.GetTarget()
+	args := call.GetArgs()
+	fnName := call.GetFunction()
+
+	var sum CostEstimate
+
+	// Traverse arguments.
+	argTypes := make([]Expr, len(args))
+	for i, arg := range args {
+		sum = sum.Add(c.cost(arg))
+		argTypes[i] = expr{t: c.types[arg.Id], expr: arg}
+	}
+
+	// Regular static call with simple name.
+	if target == nil {
+		// Check for the existence of the function.
+		fn := c.env.LookupFunction(fnName)
+		if fn == nil {
+			// not found
+			return CostEstimate{}
+		}
+		// Overwrite the function name with its fully qualified resolved name.
+		call.Function = fn.GetName()
+		sum = sum.Add(c.functionCost(call.Function, nil, argTypes))
+		return sum
+	}
+
+	// If a receiver 'target' is present, it may either be a receiver function, or a namespaced
+	// function, but not both. Given a.b.c() either a.b.c is a function or c is a function with
+	// target a.b.
+	//
+	// Check whether the target is a namespaced function name.
+	qualifiedPrefix, maybeQualified := containers.ToQualifiedName(target)
+	if maybeQualified {
+		maybeQualifiedName := qualifiedPrefix + "." + fnName
+		fn := c.env.LookupFunction(maybeQualifiedName)
+		if fn != nil {
+			// The function name is namespaced and so preserving the target operand would
+			// be an inaccurate representation of the desired evaluation behavior.
+			// Overwrite with fully-qualified resolved function name sans receiver target.
+			call.Target = nil
+			call.Function = fn.GetName()
+			return sum
+		}
+	}
+
+	// Regular instance call.
+	var targetType Expr
+	if call.Target != nil {
+		sum = sum.Add(c.cost(call.Target))
+		targetType = expr{t: c.types[call.Target.Id], expr: call.Target}
+	}
+	fn := c.env.LookupFunction(fnName)
+	// Function found, attempt overload resolution.
+	if fn != nil {
+		sum = sum.Add(c.functionCost(call.Function, &targetType, argTypes))
+		return sum
+	}
+	// Function name not declared.
+	return CostEstimate{}
+}
+
+func (c *coster) costCreateList(e *exprpb.Expr) CostEstimate {
+	create := e.GetListExpr()
+	var elemType *exprpb.Type
+	var sum CostEstimate
+	for _, e := range create.Elements {
```

Comment:
I feel like you need to either compute the cost in the aggregate or per-element. The per-element approach is likely to be more accurate in many cases as the use case for scaling expression cost by list or map size is usually when using a list or map variable whose size cannot be computed statically.

Reply:
Is createList only for when a list is declared inline in CEL, e.g. `[1, 2, 3 + 5 / 2]`? If so I think the per-element approach is right.